enhancements/installer/telemeter-metrics.md

```diff
@@ -0,0 +1,430 @@
+
+The proposed metrics fall into three categories:
+
+#### Commands
+
+##### cluster_installation_create 
+`cluster_installation_create` represents an invocation of `openshift-install create cluster`. The values for the labels would be collected throughout the execution of the command and a single sample would be pushed per installer run:
+
+```
+# HELP cluster_installation_create represents a single run of create cluster by the OpenShift installer.
+# TYPE cluster_installation_create histogram
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="15"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="20"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="25"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="30"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="35"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="40"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="45"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="50"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="55"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="60"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="+Inf"} 1
+cluster_installation_create_sum{platform="aws",result="success",version="4.2",os="linux"} 30
+cluster_installation_create_count{platform="aws",result="success",version="4.2",os="linux"} 1
+```
+The example above represents a single sample. This sample indicates that a Linux user successfully installed a 4.2 AWS cluster, which took between 30 and 35 minutes to complete installation. 
+
+###### Label Values
+* `result` - provides limited context to the outcome of an attempted install:
+    - `ProvisioningFailed`
+    - `BootstrapFailed`
+    - `APIFailed`
+    - `OperatorsDegraded`
+    - `Success`
+* `platform` - currently we deploy to 8 platforms: `aws`, `azure`, `gcp,` `libvirt`, `none`, `openstack`, `ovirt`, and `vsphere`. The list will grow. 
+* `os` - You can find all the available combinations of [Go OS and arch values](https://github.com/golang/go/blob/master/src/go/build/syslist.go) with `go tool dist list`. At this stage I assume we only want to colllect operating systems and some of the values are unlikely/impossible:
+  - `linux`
+  - `darwin` - this could be rewritten to `macos` on the installer client if we prefer
+  - `bsd` - this would encompass the `freebsd`, `netbsd`, and `openbsd` values
+  - `windows` - future proofing? 
+  - `other` - everything else not captured above
```

Comment:
Why not just use GOOS directly?  I don't see a need to invent our own, separate naming system.

Reply:
> Why not just use GOOS directly? 

The reasoning is to enforce a reduced cardinality. As we would practically only hit `linux` or `darwin` this reasoning may be flawed/over optimization. On the other hand this might be simple defensive programming.

Reply:
> The reasoning is to enforce a reduced cardinality.

Cardinality is coming up in a number of these threads, but I'm not clear on the constraint.  There is the cardinality caution [here][1], but... folks can't be running the installer _that_ often, right?  And we already have running clusters reporting metrics with much greater cardinality.  Can we accept a higher cardinality for these installer metrics and then have a window (a week?  Two?) after which we erase old data?  It's not clear to me what sort of load an increased cardinality puts on our metrics pipelines; possibly different loads for any processing that happens at Telemetry-upload time vs. query-processing time.  It's also not clear to me how much capacity our current metrics pipelines have.  You calculate cardinality in a number of places in this proposal; is there a cap we need to stay under?  Do two metrics with cardinality A have the same load as a single metric with cardinality 2A?  Feel free to say "trust me, it's really serious to keep our cardinality low, and collapsing GOOS into fewer buckets is the only only way" without trying to talk me through it ;).

[1]: https://prometheus.io/docs/practices/naming/#labels

Reply:
I appreciate your raising these points, as I have similar questions. I know that cardinality is a valid concern, but I don't know the actual capacity. Guidelines that I see leave it at cardinality should be bound ( then I look in telemter and see unbound cardinality :P).  Luckily the @openshift/openshift-team-monitoring has been giving this PR some attention and I think they can help point in the right direction. 

I do expect the installer will be considered a special case, but I think we are also abusing label so I am trying to keep this in mind.

The way I think about cardinality (I'm saying this to expose my logic/get confirmation at the risk of explaining the obvious) is that each time a metric is pushed with a unique combination of label key-value pairs, a time series is created. So we run the risk of having one-off time series created. Let's say there's a single user of I dunno `freebsd` that pushes an install metric, we will have an unchanging timeseries with a value of 1 monitored/scraped by prometheus indefinitely.

Assuming the formula I used is correct, my initial metrics proposal led to cardinality calculations over 200K. That just felt high so I was able to reorganize the metrics without losing much or any of the original data that would have been collected.

Reply:
@patrickdillon I think you understood the problem pretty well.

> then I look in telemter and see unbound cardinality

Where do you see this? The biggest point about all of this is our ability to somewhat plan for the expected load that these things would put on the system, as cardinality essentially directly translates into resource usage on the system (both for ingestion and querying). So even if there are theoretical unbound labels, they are bound in practice or in a way that we have planned into the capacity of the system.

Reply:
> Where do you see this? 

I didn't mean to imply there was a problem or worse. I was thinking of something like cluster ID. Your phrase "bound in practice" describes this situation well.